lib/active_merchant/billing/gateways/payflow.rb

```diff
@@ -188,6 +189,53 @@ def build_credit_card_request(action, money, credit_card, options)
           end
         end
         xml.target!
+        add_level_two_three_fields(xml, options)
+      end
+
+      def add_level_two_three_fields(xml, options)
+        if options[:level_two_fields] || options[:level_three_fields]
+          xml_doc = Nokogiri::XML.parse(xml.target!) ## so it doesn't put on an xml instruction
+          %i[level_two_fields level_three_fields].each do |fields|
+            xml = add_fields(xml_doc, options[fields]) if options[fields]
+          end
+        end
+        xml
```

Comment:
If `options[:level_two_fields]` and `options[:level_three_fields]` were omitted, then we'd jump straight down to here and return the Builder `xml` argument we were called with. In that case, we didn't convert the builder object to the string explicitly the same way we used to from `build_credit_card_request` with the final Builder `xml.target!` call. And thus the type of our return value changes for `build_credit_card_request` in the cases neither l2 or l3 data is used.

When this is called, we pretty much only care about the XML string built up by Builder so far, and want to further manipulate it via Nokogiri in the case these optional options are passed, right? Would it make more sense to then pass the result of `xml.target!` directly to this method, and then we know that whether we further modify it via Nokogiri or not, we can safely pass through our argument?

```
      def build_credit_card_request(action, money, credit_card, options)
        xml = Builder::XmlMarkup.new
        xml.tag! TRANSACTIONS[action] do
          # ...
        end
        add_level_two_three_fields(xml.target!, options)
      end

      def add_level_two_three_fields(xml_string, options)
        if options[:level_two_fields] || options[:level_three_fields]
          xml_doc = Nokogiri::XML.parse(xml_string)
          %i[level_two_fields level_three_fields].each do |fields|
            xml_string = add_fields(xml_doc, options[fields]) if options[fields]
          end
        end
        xml_string
      end
```

Reply:
As it stands, this receives the string output of `xml.target!` and returns either that string or the string as manipulated by Nokogiri, so we do end up with a string of XML either way. I think this makes a lot more sense, though!